cdap-app-fabric/src/main/java/io/cdap/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -148,30 +157,34 @@ public final RuntimeInfo run(ProgramDescriptor programDescriptor, ProgramOptions
 
 
     File tempDir = createTempDirectory(programId, runId);
-    Runnable cleanUpTask = createCleanupTask(tempDir, runner);
-    try {
-      // Get the artifact details and save it into the program options.
-      ArtifactId artifactId = programDescriptor.getArtifactId();
-      ArtifactDetail artifactDetail = getArtifactDetail(artifactId);
-      ProgramOptions runtimeProgramOptions = updateProgramOptions(artifactId, programId, options, runId);
-
-      // Take a snapshot of all the plugin artifacts used by the program
-      ProgramOptions optionsWithPlugins = createPluginSnapshot(runtimeProgramOptions, programId, tempDir,
-                                                               programDescriptor.getApplicationSpecification());
-
-      // Create and run the program
-      Program executableProgram = createProgram(cConf, runner, programDescriptor, artifactDetail, tempDir);
-      cleanUpTask = createCleanupTask(cleanUpTask, executableProgram);
-
-      RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(executableProgram, optionsWithPlugins), programId,
-                                                  cleanUpTask);
-      monitorProgram(runtimeInfo, cleanUpTask);
-      return runtimeInfo;
-    } catch (Exception e) {
-      cleanUpTask.run();
-      LOG.error("Exception while trying to run program", e);
-      throw Throwables.propagate(e);
-    }
+    AtomicReference<Runnable> cleanUpTaskRef = new AtomicReference<>(createCleanupTask(tempDir, runner));
+    DelayedProgramController controller = new DelayedProgramController(programRunId);
+    RuntimeInfo runtimeInfo = createRuntimeInfo(controller, programId, () -> cleanUpTaskRef.get().run());
+    monitorProgram(runtimeInfo, () -> cleanUpTaskRef.get().run());
+
+    executor.execute(() -> {
+      try {
+        // Get the artifact details and save it into the program options.
+        ArtifactId artifactId = programDescriptor.getArtifactId();
+        ArtifactDetail artifactDetail = getArtifactDetail(artifactId);
+        ProgramOptions runtimeProgramOptions = updateProgramOptions(artifactId, programId, options, runId);
+
+        // Take a snapshot of all the plugin artifacts used by the program
+        ProgramOptions optionsWithPlugins = createPluginSnapshot(runtimeProgramOptions, programId, tempDir,
+                                                                 programDescriptor.getApplicationSpecification());
+
+        // Create and run the program
+        Program executableProgram = createProgram(cConf, runner, programDescriptor, artifactDetail, tempDir);
+        cleanUpTaskRef.set(createCleanupTask(cleanUpTaskRef.get(), executableProgram));
+
+        controller.setProgramController(runner.run(executableProgram, optionsWithPlugins));
+      } catch (Exception e) {
+        controller.failed(e);
+        programStateWriter.error(programRunId, e);
+        LOG.error("Exception while trying to run program", e);
```

Comment:
Am going on a limb here, an we be more descriptive in what error happened here and what user should or can do to fix it.

Reply:
This is a systematic error that almost nothing the user can do to fix.

Reply:
It is logged for CDAP developer to debug, rather than for the user.